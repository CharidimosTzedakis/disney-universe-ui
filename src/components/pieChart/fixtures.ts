export const pieChartDataPoints = [
  {
    name: "Mikey",
    y: 2,
    films: ["Terminator", "Back to the future"],
  },
  {
    name: "Donald",
    y: 4,
    films: ["Blade runner", "film2", "film3", "film4"],
  },
  {
    name: "Taz",
    y: 7,
    films: ["film5", "film6", "film7", "film8", "film9", "film10", "film11"],
  },
  { name: "char1", y: 0, films: [] },
];
