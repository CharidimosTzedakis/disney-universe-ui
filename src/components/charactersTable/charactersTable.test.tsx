import { expect, describe, it, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "urql";
import useSearchStore from "@stores/searchStore";
import CharactersTable from "./index";
import { fromValue } from "wonka";
import { characterItems } from "@test/fixtures";

describe("<CharactersTable />", () => {
  it("renders correctly", () => {
    const fetchingState = {
      executeQuery: () =>
        fromValue({
          fetching: false,
          data: {
            characters: {
              items: characterItems,
            },
          },
        }),
    };

    const { container } = render(
      <Provider value={fetchingState}>
        <CharactersTable onCharacterSelect={() => {}} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it("it invokes callback for rendering modal when a row is selected", async () => {
    const onCharacterSelect = vi.fn();
    const fetchingState = {
      executeQuery: () =>
        fromValue({
          fetching: false,
          data: {
            characters: {
              items: characterItems,
            },
          },
        }),
    };

    const { container } = render(
      <Provider value={fetchingState}>
        <CharactersTable onCharacterSelect={onCharacterSelect} />
      </Provider>,
    );

    const row = container.querySelector('[data-row-key="112"]');
    await userEvent.click(row!);
    expect(onCharacterSelect).toHaveBeenCalledTimes(1);
    expect(onCharacterSelect).toHaveBeenCalledWith({
      id: 112,
      name: "Achilles",
    });
  });

  it("renders correctly the search results", () => {
    const fetchingState = {
      executeQuery: () =>
        fromValue({
          fetching: false,
          data: {
            characters: {
              items: [characterItems[0], characterItems[1]],
            },
          },
        }),
    };
    useSearchStore.setState({ searchMode: true });

    const { container } = render(
      <Provider value={fetchingState}>
        <CharactersTable onCharacterSelect={() => {}} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  describe("pagination", () => {
    it("navigates to next page", async () => {
      const fetchingState = {
        executeQuery: vi.fn(() =>
          fromValue({
            fetching: false,
            data: {
              characters: {
                items: [...characterItems],
              },
            },
          }),
        ),
      };

      render(
        <Provider value={fetchingState}>
          <CharactersTable onCharacterSelect={() => {}} />
        </Provider>,
      );

      const nextPageButton = await screen.findByTitle("Next Page");
      await userEvent.click(nextPageButton);

      expect(fetchingState.executeQuery).toHaveBeenCalledTimes(2);
      expect(fetchingState.executeQuery).toHaveBeenNthCalledWith(
        2,
        expect.objectContaining({
          variables: {
            filter: {},
            page: 2,
            pageSize: 50,
          },
        }),
        {},
      );
    });

    it("navigates to next page", () => {});
    it("navigates to a specific page", () => {});
  });

  describe("external state interactions", () => {
    it("updates the films pie chart data", () => {});
    it("updates the searchStore with loading state when is fetching data", () => {});
  });
  describe("data fetching", () => {
    it("fetches a page of the character table data", () => {});
    it("fetches character data based on the search filters", () => {});
  });
});
