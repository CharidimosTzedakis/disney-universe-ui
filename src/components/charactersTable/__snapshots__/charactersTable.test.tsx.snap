// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CharactersTable /> > renders correctly 1`] = `
<div>
  <div
    class="ant-table-wrapper characters-table css-dev-only-do-not-override-142vneq"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-142vneq"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table css-dev-only-do-not-override-142vneq ant-table-fixed-header ant-table-has-fix-left"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-header"
              style="overflow: hidden;"
            >
              <table
                style="table-layout: fixed; visibility: hidden;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      aria-label="Character name"
                      class="ant-table-cell ant-table-cell-fix-left ant-table-column-has-sorters"
                      scope="col"
                      style="position: sticky; left: 0px;"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Character name
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      scope="col"
                      style="position: sticky; left: 0px;"
                    >
                      # TV shows
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      # Video games
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Allies names
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Enemies names
                    </th>
                  </tr>
                </thead>
              </table>
            </div>
            <div
              class="ant-table-body"
              style="overflow-y: scroll; max-height: 700px;"
            >
              <table
                style="table-layout: fixed;"
              >
                <colgroup>
                  <col
                    style="width: 140px;"
                  />
                  <col
                    style="width: 80px;"
                  />
                  <col
                    style="width: 80px;"
                  />
                  <col
                    style="width: 100px;"
                  />
                  <col
                    style="width: 100px;"
                  />
                </colgroup>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    aria-hidden="true"
                    class="ant-table-measure-row"
                    style="height: 0px; font-size: 0px;"
                  >
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="112"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Achilles
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="18"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Abigail the Cow
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="16"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Abdullah
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="45"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Admiral Boom and Mr. Binnacle
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="7"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      .GIFfany
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="12"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      90's Adventure Bear
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="36"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Candace Adams
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="139"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Ahadi
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="152"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Al Muddy Sultan
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="181"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Irwina Allen
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          class="ant-pagination ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-142vneq"
        >
          <li
            class="ant-pagination-total-text"
          >
            1-50 of 7438 items
          </li>
          <li
            aria-disabled="true"
            class="ant-pagination-prev ant-pagination-disabled"
            title="Previous Page"
          >
            <button
              class="ant-pagination-item-link"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="left"
                class="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            tabindex="0"
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-2"
            tabindex="0"
            title="2"
          >
            <a
              rel="nofollow"
            >
              2
            </a>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-3"
            tabindex="0"
            title="3"
          >
            <a
              rel="nofollow"
            >
              3
            </a>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-4"
            tabindex="0"
            title="4"
          >
            <a
              rel="nofollow"
            >
              4
            </a>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-5 ant-pagination-item-before-jump-next"
            tabindex="0"
            title="5"
          >
            <a
              rel="nofollow"
            >
              5
            </a>
          </li>
          <li
            class="ant-pagination-jump-next ant-pagination-jump-next-custom-icon"
            tabindex="0"
            title="Next 5 Pages"
          >
            <a
              class="ant-pagination-item-link"
            >
              <div
                class="ant-pagination-item-container"
              >
                <span
                  aria-label="double-right"
                  class="anticon anticon-double-right ant-pagination-item-link-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="double-right"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 00188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 00492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
                    />
                  </svg>
                </span>
                <span
                  class="ant-pagination-item-ellipsis"
                >
                  •••
                </span>
              </div>
            </a>
          </li>
          <li
            class="ant-pagination-item ant-pagination-item-149"
            tabindex="0"
            title="149"
          >
            <a
              rel="nofollow"
            >
              149
            </a>
          </li>
          <li
            aria-disabled="false"
            class="ant-pagination-next"
            tabindex="0"
            title="Next Page"
          >
            <button
              class="ant-pagination-item-link"
              tabindex="-1"
              type="button"
            >
              <span
                aria-label="right"
                class="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            class="ant-pagination-options"
          >
            <div
              aria-label="Page Size"
              class="ant-select ant-select-outlined ant-pagination-options-size-changer css-dev-only-do-not-override-142vneq ant-select-single ant-select-show-arrow ant-select-show-search"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-wrap"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-label="Page Size"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      role="combobox"
                      type="search"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-item"
                    title="50 / page"
                  >
                    50 / page
                  </span>
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
            <div
              class="ant-pagination-options-quick-jumper"
            >
              Go to
              <input
                aria-label="Page"
                type="text"
                value=""
              />
              Page
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CharactersTable /> > renders correctly the search results 1`] = `
<div>
  <div
    class="ant-table-wrapper characters-table css-dev-only-do-not-override-142vneq"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-142vneq"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table css-dev-only-do-not-override-142vneq ant-table-fixed-header ant-table-has-fix-left"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-header"
              style="overflow: hidden;"
            >
              <table
                style="table-layout: fixed; visibility: hidden;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      aria-label="Character name"
                      class="ant-table-cell ant-table-cell-fix-left ant-table-column-has-sorters"
                      scope="col"
                      style="position: sticky; left: 0px;"
                      tabindex="0"
                    >
                      <div
                        class="ant-table-column-sorters"
                      >
                        <span
                          class="ant-table-column-title"
                        >
                          Character name
                        </span>
                        <span
                          class="ant-table-column-sorter ant-table-column-sorter-full"
                        >
                          <span
                            aria-hidden="true"
                            class="ant-table-column-sorter-inner"
                          >
                            <span
                              aria-label="caret-up"
                              class="anticon anticon-caret-up ant-table-column-sorter-up"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-up"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                />
                              </svg>
                            </span>
                            <span
                              aria-label="caret-down"
                              class="anticon anticon-caret-down ant-table-column-sorter-down"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="caret-down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                viewBox="0 0 1024 1024"
                                width="1em"
                              >
                                <path
                                  d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      scope="col"
                      style="position: sticky; left: 0px;"
                    >
                      # TV shows
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      # Video games
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Allies names
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Enemies names
                    </th>
                  </tr>
                </thead>
              </table>
            </div>
            <div
              class="ant-table-body"
              style="overflow-y: scroll; max-height: 700px;"
            >
              <table
                style="table-layout: fixed;"
              >
                <colgroup>
                  <col
                    style="width: 140px;"
                  />
                  <col
                    style="width: 80px;"
                  />
                  <col
                    style="width: 80px;"
                  />
                  <col
                    style="width: 100px;"
                  />
                  <col
                    style="width: 100px;"
                  />
                </colgroup>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    aria-hidden="true"
                    class="ant-table-measure-row"
                    style="height: 0px; font-size: 0px;"
                  >
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                    <td
                      style="padding: 0px; border: 0px; height: 0px;"
                    >
                      <div
                        style="height: 0px; overflow: hidden;"
                      >
                         
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="112"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Achilles
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      1
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="18"
                  >
                    <td
                      class="ant-table-cell ant-table-cell-fix-left"
                      style="position: sticky; left: 0px;"
                    >
                      Abigail the Cow
                    </td>
                    <td
                      class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                      style="position: sticky; left: 0px;"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      0
                    </td>
                    <td
                      class="ant-table-cell"
                    />
                    <td
                      class="ant-table-cell"
                    />
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
